src/uthash.h

```diff
@@ -365,37 +432,40 @@ do {
 do {                                                                             \
   struct UT_hash_handle *_hd_hh_del;                                             \
-  if (((delptr)->hh.prev == NULL) && ((delptr)->hh.next == NULL)) {              \
+  if (((delhh)->prev == NULL) && ((delhh)->next == NULL)) {                      \
     HASH_BLOOM_FREE((head)->hh.tbl);                                             \
     uthash_free((head)->hh.tbl->buckets,                                         \
                 (head)->hh.tbl->num_buckets * sizeof(struct UT_hash_bucket));    \
     uthash_free((head)->hh.tbl, sizeof(UT_hash_table));                          \
     (head) = NULL;                                                               \
   } else {                                                                       \
     unsigned _hd_bkt;                                                            \
-    _hd_hh_del = &((delptr)->hh);                                                \
-    if ((delptr) == ELMT_FROM_HH((head)->hh.tbl, (head)->hh.tbl->tail)) {        \
-      (head)->hh.tbl->tail = HH_FROM_ELMT((head)->hh.tbl, (delptr)->hh.prev);    \
+    _hd_hh_del = (delhh);                                                          \
+    if (_hd_hh_del == (head)->hh.tbl->tail) {                                    \
+      (head)->hh.tbl->tail = HH_FROM_ELMT((head)->hh.tbl, _hd_hh_del->prev);     \
     }                                                                            \
-    if ((delptr)->hh.prev != NULL) {                                             \
-      HH_FROM_ELMT((head)->hh.tbl, (delptr)->hh.prev)->next = (delptr)->hh.next; \
+    if (_hd_hh_del->prev != NULL) {                                              \
+      HH_FROM_ELMT((head)->hh.tbl, _hd_hh_del->prev)->next = _hd_hh_del->next;   \
     } else {                                                                     \
-      DECLTYPE_ASSIGN(head, (delptr)->hh.next);                                  \
+      DECLTYPE_ASSIGN(head, _hd_hh_del->next);                                   \
     }                                                                            \
     if (_hd_hh_del->next != NULL) {                                              \
       HH_FROM_ELMT((head)->hh.tbl, _hd_hh_del->next)->prev = _hd_hh_del->prev;   \
     }                                                                            \
-    HASH_TO_BKT(_hd_hh_del->hashv, (head)->hh.tbl->num_buckets, _hd_bkt);        \
-    HASH_DEL_IN_BKT(hh, (head)->hh.tbl->buckets[_hd_bkt], _hd_hh_del);           \
+    if (from_bucket) {                                                           \
+      HASH_TO_BKT(_hd_hh_del->hashv, (head)->hh.tbl->num_buckets, _hd_bkt);      \
+      HASH_DEL_IN_BKT(hh, (head)->hh.tbl->buckets[_hd_bkt], _hd_hh_del);         \
+    }                                                                            \
```

Comment:
You can probably guess I'm not thrilled with this construct, either. I see that `from_bucket` is always `0` or `1`, so this expands to either `if (0)` or `if (1)`.

What if you just left this code alone (functionally speaking; I do like your refactors above and will probably commit them to master soon) and let it run `HASH_TO_BKT` and a no-op `HASH_DEL_IN_BKT` in the failure case? If `_hd_hh_del->hh_prev` and `->hh_next` are always `NULL`, then this will be fine — and we don't really need to optimize the OOM case for speed, right?